tfhe/src/safe_deserialization.rs

```diff
@@ -15,13 +18,73 @@ use serde::Serialize;
 // deserialization error or worse, a garbage object.
 const SERIALIZATION_VERSION: &str = "0.3";
 
+/// Tells if this serialized object is versioned or not
+#[derive(Serialize, Deserialize, PartialEq, Eq)]
+enum SerializationMode {
+    /// Serialize with type versioning for backward compatibility
+    Versioned,
+    /// Directly serialize the type as it is provided
+    Direct,
+}
+
+/// Header with global metadata about the serialized object.
+#[derive(Serialize, Deserialize)]
+struct SerializationHeader {
+    mode: SerializationMode,
+    version: Cow<'static, str>,
+    name: Cow<'static, str>,
+}
+
+impl SerializationHeader {
+    /// Creates a new header for a versioned message
+    fn new_versioned<T: Named>() -> Self {
+        Self {
+            mode: SerializationMode::Versioned,
+            version: Cow::Borrowed(VERSIONING_VERSION),
+            name: Cow::Borrowed(T::NAME),
+        }
+    }
+
+    /// Checks the validity of a versioned message
+    fn check_versioned<T: Named>(&self) -> Result<(), String> {
+        if self.mode != SerializationMode::Versioned {
+            return Err(
+                "On deserialization, expected versioned type but got unversioned one".to_string(),
+            );
+        }
+
+        if self.version != VERSIONING_VERSION {
+            return Err(format!(
```

Comment:
add a comment + why it's currently like this, but in the future we will want to load old formats and update it to the new system

Reply:
done